quadruped_pympc/controllers/Controllers.py

```diff
@@ -0,0 +1,543 @@
+
+            if self.optimize_step_frequency==True:
+                from quadruped_pympc.controllers.gradient.nominal.centroidal_nmpc_gait_adaptive import Acados_NMPC_GaitAdaptive
+                self.batched_controller = Acados_NMPC_GaitAdaptive()
+
+        elif(self.controller_type ==  'sampling'):
+
+            if self.optimize_step_frequency==True:
+                from quadruped_pympc.controllers.sampling.centroidal_nmpc_jax_gait_adaptive import Sampling_MPC
+
+            else:
+                from quadruped_pympc.controllers.sampling.centroidal_nmpc_jax import Sampling_MPC
+
+            import jax
+            import jax.numpy as jnp
+            num_parallel_computations = self.num_parallel_computations
+            # self.iteration = self.iteration
+            self.controller = Sampling_MPC(horizon = self.horizon, 
+                                      dt = self.mpc_dt, 
+                                      num_parallel_computations = num_parallel_computations,
+                                      sampling_method = self.sampling_method,
+                                      control_parametrization = self.control_parametrization,
+                                      device="gpu")
+            self.best_control_parameters = jnp.zeros((self.controller.num_control_parameters, ))
+            self.jitted_compute_control = jax.jit(self.controller.compute_control, device=self.controller.device)
+            # jitted_get_key = jax.jit(controller.get_key, device=controller.device)
+            self.jitted_prepare_state_and_reference = self.controller.prepare_state_and_reference
+
+            self.index_shift = 0
+
+        # Tracking mpc error (z, x_dot, y_dot, z_dot, roll, pitch, roll_dot, pitch_dot, yaw_dot)
+        self.mean_tracking = np.zeros((9,))
+        self.mean_num_sample = 1
+        self.mean_optimizer_cost = 0
+        self.optimizer_cost = 0
+
+
+
+
+        
+    def periodic_gait_generator(self):
+
+        # Periodic gait generator _________________________________________________________________________
+        gait_name = self.gait_name
+        gait_type, duty_factor, step_frequency = self.get_gait_params(gait_name)
+        #
+        self.duty_factor = duty_factor
+        self.step_frequency = step_frequency
+        self.gait_type = gait_type
+        self.pgg = PeriodicGaitGenerator(duty_factor=duty_factor, step_freq=step_frequency, gait_type=self.gait_type,
+                                        horizon=self.horizon * 2, contact_sequence_dt=self.mpc_dt/2.)
+        
+        self.contact_sequence = self.pgg.compute_contact_sequence()
+        self.nominal_sample_freq = self.step_frequency
+        # compute stance and swing tme based on gate selection!
+        self.stance_time = (1 / self.step_frequency) * self.duty_factor
+        self.swing_period = (1 - self.duty_factor) * (1 / self.step_frequency)  # + 0.07
+
+    
+
+    def initialize(self):
```

Comment:
This code should be in the class initializer. Why define a "initialize" function?